src/s3fs.cpp

```diff
@@ -3148,141 +3086,25 @@ static int s3fs_opendir(const char* _path, struct fuse_file_info* fi)
-{
-    if(!s3fscurl){
-        return nullptr;
-    }
-    size_t ssec_key_pos= s3fscurl->GetLastPreHeadSeecKeyPos();
-    int retry_count = s3fscurl->GetMultipartRetryCount();
-
-    // retry next sse key.
-    // if end of sse key, set retry master count is up.
-    ssec_key_pos = (ssec_key_pos == static_cast<size_t>(-1) ? 0 : ssec_key_pos + 1);
-    if(0 == S3fsCurl::GetSseKeyCount() || S3fsCurl::GetSseKeyCount() <= ssec_key_pos){
-        if(s3fscurl->IsOverMultipartRetryCount()){
-            S3FS_PRN_ERR("Over retry count(%d) limit(%s).", s3fscurl->GetMultipartRetryCount(), s3fscurl->GetSpecialSavedPath().c_str());
-            return nullptr;
-        }
-        ssec_key_pos = -1;
-        retry_count++;
-    }
-
-    std::unique_ptr<S3fsCurl> newcurl(new S3fsCurl(s3fscurl->IsUseAhbe()));
-    const std::string& path       = s3fscurl->GetBasePath();
-    const std::string& base_path  = s3fscurl->GetBasePath();
-    const std::string& saved_path = s3fscurl->GetSpecialSavedPath();
-
-    if(!newcurl->PreHeadRequest(path, base_path, saved_path, ssec_key_pos)){
-        S3FS_PRN_ERR("Could not duplicate curl object(%s).", saved_path.c_str());
-        return nullptr;
-    }
-    newcurl->SetMultipartRetryCount(retry_count);
-
-    return newcurl;
-}
-
 static int readdir_multi_head(const char* path, const S3ObjList& head, void* buf, fuse_fill_dir_t filler)
 {
-    // [TODO]
-    // This will be checked and changed after removing S3fsMultiCurl
-    //
-    S3fsMultiCurl curlmulti(S3fsCurl::GetMaxMultiRequest(), true);      // [NOTE] run all requests to completion even if some requests fail.
-    s3obj_list_t  headlist;
-    int           result = 0;
-
-    S3FS_PRN_INFO1("[path=%s][list=%zu]", path, headlist.size());
+    S3FS_PRN_INFO1("[path=%s][head=<%s>][filler=%p]", path, head.IsEmpty() ? "empty" : "not empty", filler);
 
     // Make base path list.
+    s3obj_list_t  headlist;
     head.GetNameList(headlist, true, false);                                        // get name with "/".
     StatCache::getStatCacheData()->GetNotruncateCache(std::string(path), headlist); // Add notruncate file name from stat cache
 
-    // Initialize S3fsMultiCurl
-    curlmulti.SetSuccessCallback(multi_head_callback);
-    curlmulti.SetRetryCallback(multi_head_retry_callback);
-
-    // Success Callback function parameter(SyncFiller object)
+    // Initialize SyncFiller object
     SyncFiller syncfiller(buf, filler);
-    curlmulti.SetSuccessCallbackParam(reinterpret_cast<void*>(&syncfiller));
 
-    // Not found Callback function parameter
-    struct multi_head_notfound_callback_param notfound_param;
-    if(support_compat_dir){
-        curlmulti.SetNotFoundCallback(multi_head_notfound_callback);
-        curlmulti.SetNotFoundCallbackParam(reinterpret_cast<void*>(&notfound_param));
-    }
+    // common variables
+    Semaphore    multi_head_sem(0);
+    int          req_count  = 0;
+    int          req_result = 0;
```

Comment:
I think this is probably better expressed as a `std::future` but we can address this in a separate commit.

Reply:
Thanks.
I'll fix it after I merge the PRs.